L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc

```diff
@@ -0,0 +1,905 @@
+GlobalPoint HGCalTriggerGeometryV9Imp3::getModulePosition(const unsigned module_det_id) const {
+  unsigned det = DetId(module_det_id).det();
+  // Position: barycenter of the module.
+  Basic3DVector<float> moduleVector(0., 0., 0.);
+  const auto cell_ids = getCellsFromModule(module_det_id);
+  // Scintillator
+  if (det == DetId::HGCalHSc) {
+    for (const auto& cell : cell_ids) {
+      moduleVector += hscGeometry()->getPosition(cell).basicVector();
+    }
+  }
+  // HFNose
+  else if (det == DetId::Forward && DetId(module_det_id).subdetId() == ForwardSubdetector::HFNose) {
+    for (const auto& cell : cell_ids) {
+      HFNoseDetId cellDetId(cell);
+      moduleVector += noseGeometry()->getPosition(cellDetId).basicVector();
+    }
+  }  // Silicon
+  else {
+    for (const auto& cell : cell_ids) {
+      HGCSiliconDetId cellDetId(cell);
+      moduleVector += (cellDetId.det() == DetId::HGCalEE ? eeGeometry()->getPosition(cellDetId)
+                                                         : hsiGeometry()->getPosition(cellDetId))
+                          .basicVector();
+    }
+  }
+
+  return GlobalPoint(moduleVector / cell_ids.size());
+}
+
+void HGCalTriggerGeometryV9Imp3::fillMaps() {
+  // read module mapping file
+  std::ifstream l1tModulesMappingStream(l1tModulesMapping_.fullPath());
+  if (!l1tModulesMappingStream.is_open()) {
+    throw cms::Exception("MissingDataFile") << "Cannot open HGCalTriggerGeometry L1TModulesMapping file\n";
+  }
+
+  short waferu = 0;
+  short waferv = 0;
+  short module = 0;
+  short layer = 0;
+  for (; l1tModulesMappingStream >> layer >> waferu >> waferv >> module;) {
+    wafer_to_module_.emplace(packLayerWaferId(layer, waferu, waferv), module);
+    module_to_wafers_.emplace(packLayerModuleId(layer, module), packWaferId(waferu, waferv));
+  }
+  if (!l1tModulesMappingStream.eof()) {
+    throw cms::Exception("BadGeometryFile")
+        << "Error reading L1TModulesMapping '" << layer << " " << waferu << " " << waferv << " " << module << "' \n";
+  }
+  l1tModulesMappingStream.close();
+  // read links mapping file
+  std::ifstream l1tLinksMappingStream(l1tLinksMapping_.fullPath());
+  if (!l1tLinksMappingStream.is_open()) {
+    throw cms::Exception("MissingDataFile") << "Cannot open HGCalTriggerGeometry L1TLinksMapping file\n";
+  }
+  short links = 0;
+  const short max_modules_60deg_sector = 127;
+  for (; l1tLinksMappingStream >> layer >> module >> links;) {
+    if (module_to_wafers_.find(packLayerModuleId(layer, module)) == module_to_wafers_.end()) {
+      links = 0;
+    }
+    if (module > max_modules_60deg_sector)
+      sector0_mask_ = 0xff;  // Use 8 bits to encode module number in 120deg sector
+    links_per_module_.emplace(packLayerModuleId(layer, module), links);
+  }
+  if (!l1tLinksMappingStream.eof()) {
+    throw cms::Exception("BadGeometryFile")
+        << "Error reading L1TLinksMapping '" << layer << " " << module << " " << links << "' \n";
+  }
+  l1tLinksMappingStream.close();
```

Comment:
All these mapping filling should be replaced. All these information should in principle be retrieved from the json file.